README.pod
lib/MR/APNS.pm

Changes
Revision history for Perl extension MR::APNS.

0.01  Fri Mar 15 14:01:24 2013
	- original version; created by h2xs 1.23 with options
		-AX MR::APNS



lib/MR/APNS/Payload.pm
package MR::APNS::Payload;
use Mouse;
use JSON::XS qw();
use Encode qw();
use bytes;

with qw(
    MR::APNS::Role::PayloadError
    MR::APNS::Role::PayloadAction
);

# you can set either token or bintoken
has 'token'  => (
    is              => 'ro',
    isa             => 'Str',
    required        => 1 
);

has 'bintoken'  => (
    is              => 'ro',
    isa             => 'Str',
    required        => 1 
);

has 'alert'  => (
    is              => 'rw',
    isa             => 'Str|HashRef',
    predicate       => 'has_alert'
);

around 'alert' => sub {
    my $orig = shift;
    my $self = shift;
    return $self->$orig(@_) unless (@_ and $self->_utf8_on);
    
    my $alert = @_ == 1 ? $_[0] : {@_};
    unless (ref $alert) {
        Encode::_utf8_on($alert);
    } else {
        Encode::_utf8_on($alert->{body}) if exists $alert->{body};
        if (exists $alert->{'loc-args'}) {
            Encode::_utf8_on($_) for @{$alert->{'loc-args'}};
        } 
    }
    return $self->$orig($alert);
};

has 'badge'  => (
    is              => 'rw',
    isa             => 'Int',
    predicate       => 'has_badge'
);

has 'sound'  => (
    is              => 'rw',
    isa             => 'Str',
    predicate       => 'has_sound'
);

has 'custom'  => (
    is              => 'rw',
    isa             => 'Maybe[HashRef]',
    clearer         => 'clean_custom',
    documentation   => 'any additional dictory'
);

has 'expiry'  => (
    is              => 'rw',
    isa             => 'Int',
    lazy            => 1,
    default         => sub { 0 },
    documentation   => 'Date expressed in seconds (UTC) that identifies when the notification is no longer valid and can be discarded'
);

has 'error'  => (
    is              => 'rw',
    isa             => 'PayloadError',
    lazy            => 1,
    default         => sub { -1 },
    documentation   => 'error code. see also MR::APNS::Role::PayloadError'
);

has 'context'  => (
    is              => 'rw',
    isa             => 'Any',
    lazy            => 1,
    default         => sub { undef },
    documentation   => 'anything that can help to relate this instance with another code'    
);

has '_can_skip_custom'  => (
    is              => 'rw',
    isa             => 'Bool',
    lazy            => 1,
    default         => sub { 0 },
    documentation   => 'try to skip custom when payload is too big'
);

has '_identifier'  => (
    is              => 'ro',
    isa             => 'Int',
    lazy            => 1,
    default         => sub { hex(["$_[0]"=~/\((0x[0-9a-f]+)\)/]->[0]) },
    documentation   => 'This same identifier is returned in a error-response packet if APNs cannot interpret a notification.'
);

has '_utf8_on'  => (
    is              => 'ro',
    isa             => 'Bool',
    required        => 1,
    default         => sub { 0 },
    documentation   => 'utf8 flag up for alert (simple alert, or "body" and "loc-args" keys)'
);

has 'as_binary'  => (
    is              => 'rw',
    isa             => 'Maybe[Str]',
    lazy            => 1,
    default         => sub {
        my $self = shift;
        my $json = $self->custom || {};
        
        my $not_empty = 0;
        for ( qw(alert badge sound) ) {
            my $predicate = 'has_'.$_;
            if ($self->$predicate) {
                $json->{aps}->{$_} = $self->$_;
                $not_empty ||= 1;
            }
        }
        return unless $not_empty;
        
        $json = JSON::XS->new->utf8->encode($json);
        if (bytes::length $json > 256) {
            if ($self->custom and $self->_can_skip_custom) {
                $self->clean_custom;
                return $self->to_bin;
            } else {
                $self->error(-7);
                return;
            }
        }
        
        return pack(
            'C L N n/a* n/a*',
            1,
            $self->_identifier,
            $self->expiry,
            $self->bintoken,
            $json
        );        
    },
    documentation   => 'payload to send'
);

sub BUILDARGS {
    my $class = shift;
    my $param = @_==1 ? $_[0] : {@_};
    if ($param->{token} and not $param->{bintoken}) {
        $param->{bintoken} = pack('H*', $param->{token});
    } elsif ($param->{bintoken} and not $param->{token}) {
        $param->{token} = unpack('H*', $param->{bintoken});
    };
    return $param; 
}

sub BUILD {
    my $self = shift;
    # utf8 up
    $self->alert($self->alert) if $self->_utf8_on;
    return $self;
}

__PACKAGE__->meta->make_immutable;
no Mouse;
1;
__END__
=head1 NAME

MR::APNS::Payload - push notification instance

=head1 DESCRIPTION

It's L<Mouse> class

=head1 ATTRIBUTES

=head3 Required

=over

=item token : Str (ro)

=item bintoken : Str (ro)

It's device token. Use either of them.

=back

=head3 Optional

=over

=item alert : Str:HashRef (rw)

Message

=item badge : Int (rw)

The number to display as the badge of the application icon.
To remove the badge, set the value of this property to 0.

=item sound : Str (rw)

The name of a sound file in the application bundle. 
The sound in this file is played as an alert. 
If the sound file doesn't exist or I<default> is specified as the value, the default alert sound is played.

=item custom : HashRef (rw)

Custom dictionary which need include to payload.

=item _can_skip_custom : Bool (rw)

When payload has size more than 256 byte, it's able to skip custom dictionary. B<Default>: 0

=item expiry : Int (rw)

Date expressed in seconds (UTC) that identifies when the notification is no longer valid and can be discarded. B<Default>: 0

=item error : Int (rw)

=item error_str : Str (rw)

See source code L<MR::APNS::Role::PayloadError>. In common case use L<error_str>.

=item context : Any (rw)

Anything that can help to relate this instance with another code

=item  _identifier : Int (ro)

The first rule of C<_identifier> is: You do not talk about C<_identifier>.

=item _utf8_on : Bool (ro)

Invoke C<Encode::_utf8_on> on simple alert or "body" and "loc-args" keys.

=back

=head1 METHODS

=head2 as_binary()

Make payload

=head1 SEE ALSO and DEPENDENCIES

L<MR::APNS>

=head1 LICENSE

This library is under meow license.

=cut

lib/MR/APNS/Transport.pm
package MR::APNS::Transport;
use Mouse;

use File::Temp  qw();
use IO::Select  qw();
use Socket      qw();
use Net::SSLeay qw();
use Carp        qw();
use IO::Handle  qw();
use List::Util  qw();

has 'cert_file' => (
    is              => 'ro',
    isa             => 'Str',
    predicate       => 'has_cert_file',
    documentation   => 'cart .pem file'
);

has 'key_file' => (
    is              => 'ro',
    isa             => 'Str',
    lazy            => 1,
    default         => sub { shift->cert_file },
    documentation   => 'yes, by default it is cert_file'
);

sub _temp_pem_file {
    my ($content, $type) = @_;
    my $fh = File::Temp->new(TEMPLATE => "mr-apns-".$type."-".time()."-$$-XXXXXXXX", TEMPDIR  => 1, EXLOCK   => 0);
    syswrite $fh, $content;
    close $fh;
    return $fh->filename;
}

has 'cert'  => (
    is              => 'ro',
    isa             => 'Str',
    documentation   => 'content of cert .pem file',
    trigger => sub {
        my $self = shift;
        $self->cert_file( _temp_pem_file($_[0], 'cert') );
    } 
);

has 'key'  => (
    is              => 'ro',
    isa             => 'Str',
    documentation   => 'content of key .pem file',
    trigger => sub {
        my $self = shift;
        $self->key_file( _temp_pem_file($_[0], 'key') );
    } 
);

has 'password'  => (
    is              => 'ro',
    isa             => 'Str|CodeRef',
    documentation   => 'key password'
);

has 'sandbox'  => (
    is              => 'ro',
    isa             => 'Bool',
    lazy            => 1,
    default         => sub { 0 },
    documentation   => 'gateway target'
);

has 'feedback'  => (
    is              => 'rw',
    isa             => 'Bool',
    lazy            => 1,
    default         => sub { 0 },
    documentation   => 'feedback target'
);

has 'hostname' => (
    is              => 'ro',
    isa             => 'Str',
    lazy            => 1,
    default         => sub {
        my $self = shift;
        return sprintf('%s.%spush.apple.com', $self->feedback ? 'feedback' : 'gateway' ,$self->sandbox ? 'sandbox.' : '')
    },
);

has 'port' => (
    is              => 'ro',
    isa             => 'Int',
    lazy            => 1,
    default         => sub {
        my $self = shift;
        $self->feedback ? 2196 : 2195 
    },    
);

has ['write_timeout', 'last_read_timeout', 'feedback_timeout'] => (
    is              => 'rw',
    isa             => 'Num',
    lazy            => 1,
    default         => sub { 0.1 }
);

has '_ioselect'  => (
    is              => 'rw',
    isa             => 'Maybe[IO::Select]',
    lazy            => 1,
    predicate       => '_has_ioselect',
    clearer         => '_clean_ioselect',
    default         => sub {
        my $self = shift;
        socket(my $sock, Socket::PF_INET(), Socket::SOCK_STREAM(), 0) or die "can't create socket: $!";
        my $iaddr = Socket::inet_aton($self->hostname) or die sprintf "can't create iaddr from %s: %s", $self->hostname, $!;
        my $sock_addr = Socket::pack_sockaddr_in($self->port, $iaddr) or die "can't create sock_addr: $!";
        connect($sock, $sock_addr) or die "can't connect socket: $!";
        $sock->autoflush;
        return IO::Select->new($sock);
    },
    documentation   => 'simple select wrapper'
);

has '_ctx'  => (
    is              => 'rw',
    isa             => 'Int',
    lazy            => 1,
    predicate       => '_has_ctx',
    clearer         => '_clean_ctx',
    default         => sub {
        my $self = shift;
        my $ctx = Net::SSLeay::CTX_tlsv1_new() or _die_if_ssl_error("can't create SSL_CTX: $!");
        Net::SSLeay::CTX_set_options($ctx, Net::SSLeay::OP_ALL());
        _die_if_ssl_error("ctx options: $!");
        my $pw = $self->password;
        Net::SSLeay::CTX_set_default_passwd_cb($ctx, ref $pw ? $pw : sub { $pw });
        Net::SSLeay::CTX_use_certificate_file($ctx, $self->cert_file, Net::SSLeay::FILETYPE_PEM());
        _die_if_ssl_error("certificate: $!");
        Net::SSLeay::CTX_use_RSAPrivateKey_file($ctx, $self->cert_file, Net::SSLeay::FILETYPE_PEM());
        _die_if_ssl_error("private key: $!");        
        return $ctx;
    },
    documentation   => 'SSL_CTX object'
);

has '_ssl'  => (
    is              => 'rw',
    isa             => 'Int',
    lazy            => 1,
    predicate       => '_has_ssl',
    clearer         => '_clean_ssl',
    default         => sub {
        my $self = shift;
        my $ssl = Net::SSLeay::new($self->_ctx);
        Net::SSLeay::set_fd($ssl, fileno [$self->_ioselect->handles]->[0]);
        Net::SSLeay::connect($ssl) or _die_if_ssl_error("failed ssl connect: $!");
        return $ssl;
    },
    documentation   => 'SSL_CTX object'
);

has 'connect'  => (
    is              => 'rw',
    isa             => 'Bool',
    lazy            => 1,
    clearer         => '_clean_connect',
    predicate       => '_has_connect',
    default         => sub {
        my $self = shift;
        # start chain action
        $self->_ssl;
        return 1;
    },
    documentation   => 'connect state'
);

sub disconnect {
    my $self = shift;
    return unless $self->_has_connect;
    eval {
        my @sockets = ();
        if ($self->_has_ioselect) {
            @sockets = $self->_ioselect->handles;
            $self->_ioselect->remove(@sockets) if @sockets;
            for (@sockets) {
                die "can't shutdown socket: $!" unless defined CORE::shutdown($_, 1)
            };
        }
        
        if ($self->_has_ssl) {
            Net::SSLeay::free($self->_ssl);
            _die_if_ssl_error("failed ssl free: $!");
        }
    
        if ($self->_has_ctx) {
            Net::SSLeay::CTX_free($self->_ctx);
            _die_if_ssl_error("failed ctx free: $!");
        }
        
        close $_ or die "can't close socket: $!" for @sockets;
        
        $self->_clean_ctx; 
        $self->_clean_ssl;
        $self->_clean_ioselect;
        $self->_clean_connect;
    };
    return $@;
}

sub send {
    my $self = shift;
    my @messages = grep { $_->as_binary } @_;
    
    my $send_cnt = 0; my $error = '';

    my @apns_error = (); my $i = 0;
    for my $m (@messages) {
        my $status = $self->_send($m->as_binary, $i == @messages ? 1 : 0);
        if ($status->{error} eq 'timeout') {
            $m->error(-200);
        } elsif ($status->{error} eq 'fatal') {
            $error = $status->{error_str};
            last;
        } elsif ($status->{error} eq 'OK' and $status->{apns_error}) {
            push @apns_error, $status->{apns_error};
        }
        
        if ($status->{send}) {
            $send_cnt++;
            $m->error(-2);
        }
        
        $i++;
    }
    
    eval { $self->disconnect };
    $error = $@ if (not $error and $@);
    
    my %apns_error = map { %$_ } @apns_error;
    for (@messages) {
         if ( $apns_error{ $_->_identifier } ) {
             # ignore APNS_NO_ERRORS
             eval { $_->error($apns_error{ $_->_identifier }) };
             if ($@) {
                 $_->error(-255);
             };
         }
    }
    
    return ($send_cnt, $error);
}

sub _send {
    my $self = shift;
    my $payload = shift;
    my $last = shift;
    my $ret = {};
    eval { $self->connect };
    if ($@) {
        return {error => 'fatal', error_str => $@, send => 0};
    }
    my ($ready_socket) = $self->_ioselect->can_write( $self->write_timeout );
    unless ($ready_socket) {
        return {error => 'timeout', error_str => "socket timeout on write (0.1)", send => 0};
    }
    eval {
        Net::SSLeay::ssl_write_all($self->_ssl, $payload) or _die_if_ssl_error("ssl_write_all error: $!");
    };
    if ($@) {
        return {error => 'fatal', error_str => $@, status => 0};
    }
    ($ready_socket) = $self->_ioselect->can_read($last ? $self->last_read_timeout : 0);
    unless ($ready_socket) {
        return {error => 'OK', send => 1};
    }
    my $data = eval {
        Net::SSLeay::ssl_read_all($self->_ssl) or _die_if_ssl_error("ssl_read_all error: $!");
    };
    if ($@) {
        return {error => 'fatal', error_str => $@, send => 1};
    }
    unless ($data) {
        # apns closed connection (ignore error for payloads)
        $self->disconnect;
        return {error => 'OK', send => 1};
    };
    while ($data) {
        my ($command, $status, $identifier);
        ($command, $status, $identifier, $data) = unpack 'CCL a*', $data;
        $ret->{ $identifier } = $status;
    }    
    return {error => 'OK', send => 1, apns_error => $ret};
}

sub _die_if_ssl_error {
    my $err = Net::SSLeay::print_errs("SSL error: $_[0]");
    Carp::croak $err if $err;
}

sub retrieve_feedback {
    my $self = shift;
    my $res = {};
    $self->feedback(1);
    eval {
        $self->connect;
        my ($ready_socket) = $self->_ioselect->can_read($self->feedback_timeout);
        return unless $ready_socket;
        my $data = Net::SSLeay::ssl_read_all($self->_ssl) or _die_if_ssl_error("ssl_read_all error: $!");
        while ($data) {
            my ($time_t, $bintoken);
            ($time_t, $bintoken, $data) = unpack 'N n/a a*', $data;
            my $token = unpack 'H*', $bintoken;
            if (exists $res->{$token}) {
                $res->{$token}->{time_t} = List::Util::max($time_t, $res->{$token}->{time_t});
            } else {
                $res->{$token} = {
                    time_t      => $time_t,
                    bintoken    => $bintoken,
                    token       => $token                    
                };
            }
        }
        $self->disconnect;
    };
    return (keys %$res ? [values %$res] : undef, $@);
}

sub BUILD {
    my $self = shift;
    unless ($self->has_cert_file) {
        warn "cert_file is a necessary attribute";
        return;
    }
    Net::SSLeay::load_error_strings();
    Net::SSLeay::SSLeay_add_ssl_algorithms();
    Net::SSLeay::randomize();
    return $self;  
}

sub DEMOLISH {
    my $self = shift;
    $self->disconnect;
}

__PACKAGE__->meta->make_immutable;
no Mouse;
1;
__END__


lib/MR/APNS/Role/PayloadError.pm
package MR::APNS::Role::PayloadError;
use Mouse::Role;
use Mouse::Util::TypeConstraints;

use vars qw(
    $APNS_ERROR $PAYLOAD_ERROR $TRANSPORT_ERROR $ERROR
);

$APNS_ERROR = {
   0    => 'APNS_NO_ERRORS',
   1    => 'APNS_PROCESSING_ERROR',
   2    => 'APNS_MISSING_DEVICE_TOKEN',
   3    => 'APNS_MISSING_TOPIC',
   4    => 'APNS_MISSING_PAYLOAD',
   5    => 'APNS_INVALID_TOKEN_SIZE',
   6    => 'APNS_INVALID_TOPIC_SIZE',
   7    => 'APNS_INVALID_PAYLOAD_SIZE',
   8    => 'APNS_INVALID_TOKEN',
   255  => 'APNS_UNKNOWN_ERROR',
};

$PAYLOAD_ERROR = {
   -1   => 'PAYLOAD_CREATE',
   -2   => 'PAYLOAD_SEND',
   -7   => 'PAYLOAD_INVALID_SIZE',        
};

$TRANSPORT_ERROR = {
   -200 => 'TRANSPORT_TIMEOUT',
   -255 => 'TRANSPORT_UNKNOWN_ERROR'    
};

$ERROR = {
    %$APNS_ERROR,
    %$PAYLOAD_ERROR,
    %$TRANSPORT_ERROR
};

enum 'PayloadError'  => ( keys %$ERROR );

sub error_str {
    my $self = shift;
    return $ERROR->{ $self->error };
}

no Mouse::Util::TypeConstraints;
1;
__END__


lib/MR/APNS.pm
package MR::APNS;

our $VERSION = '0.01';

use Mouse;
use MR::APNS::Transport;
use MR::APNS::Payload;

use Exporter qw(import);
use vars     qw(@EXPORT @EXPORT_OK %EXPORT_TAGS);

@EXPORT      = qw();
@EXPORT_OK   = qw(ACTION_NONE ACTION_RESEND ACTION_DELETE);
%EXPORT_TAGS = (
    action => [qw(ACTION_NONE ACTION_RESEND ACTION_DELETE)]
);

use constant {
    ACTION_NONE     => 0,
    ACTION_RESEND   => 1,
    ACTION_DELETE   => 2
};

has 'error'  => (
    is              => 'rw',
    isa             => 'Str',
    clearer         => 'clean_error',
    lazy            => 1,
    default         => sub { '' },
    documentation   => 'connect/transport error'
);

has 'transport'  => (
    is              => 'rw',
    isa             => 'MR::APNS::Transport',
    required        => 1
);


sub send {
    my $self = shift;
    $self->clean_error();
    my @messages = ref $_[0] eq 'ARRAY' ? @{$_[0]} : @_;
    return 0 unless @messages;
    
    my ($send_cnt, $error) = $self->transport->send(@messages);
    $self->error($error) if $error;
    return $send_cnt;
}

sub retrieve_feedback {
    my $self = shift;
    $self->clean_error();
    my ($feedback, $error) = $self->transport->retrieve_feedback;
    $self->error($error) if $error;
    return $feedback;
}

sub BUILDARGS {
    my $class = shift;
    my $param = @_==1 ? $_[0] : {@_};
    $param->{transport} ||= MR::APNS::Transport->new($param);
    return $param; 
}

sub BUILD {
    my $self = shift;
    return $self->transport ? $self : undef;
}

__PACKAGE__->meta->make_immutable;
no Mouse;
1;
__END__
=encoding utf-8

=head1 NAME

MR::APNS - easy way to send push notification with enhanced format for packets

=head1 SYNOPSIS

=head3 sending push notifications

    use MR::APNS qw(:action);
    
    my $token       = "9f19404f78b8e440d2af15767dce877be3e18f81657cc8861f370fa3311a2077";
    my $bin_token   = pack 'H*', $device_token;
    
    my @m = (  
        MR::APNS::Payload->new(token => $token, alert => 'My first message'),
        MR::APNS::Payload->new(bintoken => $bin_token, alert => 'My second message')
    );
    
    my $apns = MR::APNS->new(
        sandbox => 1,
        cert_file => '/path/to/cert.pem'
    );
    
    my $send_cnt = $apns->send(@m);
    
    if ($apns->error or $send_cnt != @m) {
        for (@m) {
            my $action = $_->need_action;
            if ($action == ACTION_NONE) {
                # success
                print "message already send\n";
            } elsif ($action == ACTION_RESEND) {
                # timeout or unsended because transport error
                print "message need to resend (error: ".$_->error_str.")\n";
            } elsif ($action == ACTION_DELETE) {
                # don't try to resend
                print "message malformed (error: ".$_->error_str.")\n";
            }
        }
        # transport error
        die $apns->error if $apns->error;
    }

=head3 retrieve feedback

    use MR::APNS;
    
    my $apns = MR::APNS->new(
        sandbox => 1,
        cert_file => '/path/to/cert.pem'
    );
    
    while (1) {
        my $feedback = $apns->retrieve_feedback;
        if ($feedback) {
            for (@$feedback) {
                .. to forget about device tokens ..
                ..there're keys: time_t, bintoken, token ...
            };
        }
        die $apns->error if $apns->error;
        last unless $feedback;
    }

=head1 DESCRIPTION

It's a very gentle APNS (Apple Push Notifications Service) client with enhanced format for notification packets.

=head1 METHODS

=head2 new(%args)

create a new instance of C<MR::APNS::Transport> and C<MR::APNS>.

=head3 Required arguments are

=over

=item cert_file : file path (ro)

=item cert : Str (ro)

Either of them is required. Sets certificate. When C<cert> was specified then C<cert_file> will be ignored.

=back

=head3 Optional arguments are

=over

=item key_file : file path(ro)

=item key : Str(ro)

=item password : Str|CodeRef(ro)

Private key settings. B<Default> : is the same as C<cert_file>

=item sandbox : Bool(ro)

Gateway target. B<Default> : 0

=item hostname : Str(ro)

Gateway target. B<Default>: depends on C<sandbox>

=item port : Int(ro)

Gateway target. B<Default> : 2195

=item write_timeout : Num(rw)

It's timeout for L<IO::Select> to I<can_write>. It'll be invoked before sending every payload. B<Default>: 0.1

=item last_read_timeout : Num(rw)

It's timeout for L<IO::Select> to I<can_read>. It'll be invoked after sending last payload to receive error-response packets.

Between sending payloads I<can_read> invoke with zero timeout.  B<Default>: 0.1

=back

=head2 send(@messages)

Send notification for APNS. It take a list of L<MR::APNS::Payload> instances. Returns the number of success sended items.

=head2 retrieve_feedback()

Retrieve data from APNS feedback service. Returns an arrayref (possibly) containing hashrefs. eg:

    [
        {
            'time_t' => 1259577923,
            'token' => '04ef31c86205...624f390ea878416',
            'bintoken' => '..binary representation of token..'
        },
        ...
    ]

It's possibly to have feedback and C<error> at the same time when error happens on disconnect.

=head2 error

Last transport error as string.

=head2 transport

Read only access to C<MR::APNS::Transport> instance. C<MR::APNS::Transport> doesn't have a POD.
 
Avaliable attributes are (see above) C<cert_file>, C<cert>, C<key_file>, C<key>, C<password>, C<sandbox>, C<hostname>, C<port>, 
C<write_timeout>, C<last_read_timeout>, C<feedback> and a few methods are C<connect>, C<disconnect>, C<send>, C<get_feedback>.

In common case you don't need direct access to C<transport> 

=head1 EXPORT TAGS

=head2 :action

After sending it is possible to filter messages to receive next action need to do.

=over

=item ACTION_NONE

Message was pushed.

=item ACTION_RESEND

Transport returned timeout error or connection was interrupted.

=item ACTION_DELETE

Message is malformed

=back

=head1 MORE EXAMPLES

=head3 specify 'context' is very useful

    # 'context' has type 'Any'
    my $m = MR::APNS::Payload->new(token => $token, ..., context => $storage_id);
    
    # in common case you shouldn't to send single notification otherwise APNS will be unhappy
    unless ( $apns->send($m) ) {
        my $action = $m->need_action;
        if ($action == ACTION_RESEND) {
            print "message with storage id ".$m->context." need to resend (error: ".$m->error_str.")\n";
            ...
        } elsif ($action == ACTION_DELETE) {
            print "message with storage id ".$m->context." is malformed (error: ".$m->error_str.")\n";
        }
    };

=head3 'utf8' flag  
    
It's case when you send alert whitout utf8 flag.

    my $alert = "Привет, APNS";
    # or
    $alert = { body => $alert, loc-args => [...], ... };
    
    # device should receive correct message
    my $m = MR::APNS::Payload->new(token => $token, ..., _utf8_on => 1);

=head1 NOTE

I<cert_file> and I<key_file> are the same file if you've generated them as in example L<http://stackoverflow.com/questions/11536587/creating-pem-file-for-push-notification>.

=head1 SEE ALSO

L<MR::APNS::Payload>

=head1 ALL DEPENDENCIES

L<Mouse>, L<Mouse::Role>, L<Mouse::Util::TypeConstraints>, L<File::Temp>, L<IO::Select>, L<Net::SSLeay>, 
L<JSON::XS>, L<Encode>, L<List::Util>

=head1 LICENSE

This library is under meow license.

=cut
